docs/root/configuration/secret.rst

```diff
@@ -0,0 +1,146 @@
+.. _config_secret_discovery_service:
+
+Secret discovery service (SDS)
+==============================
+
+TLS certificates, the secrets, can be specified in the bootstrap.static_resource
+:ref:`secrets <envoy_api_field_config.bootstrap.v2.Bootstrap.StaticResources.secrets>`.
+But they can also be fetched remotely by secret discovery service (SDS).
+
+The most important benefit of SDS is to simplify the certificate management. Without this feature, in k8s deployment, certificates must be created as secrets and mounted into the proxy containers. If certificates are expired, the secrets need to be updated and the proxy containers need to be re-deployed. With SDS, a certral SDS server will push certificates to all proxy instances. If certificates are expired, the server just pushes new certificates to Envoy instances, Envoy will use the new ones right away without re-deployment.
+
+If a listener server certificates needed to be fetched by SDS remotely, it will NOT be marked as active, its port will not be opened before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the listener will be marked as active, and the port will be open, but the connection to the port will be reset.
+
+Upstream cluster is handled in the similar way, if a cluster client certificates needed to be fetched by SDS remotely, it will NOT be marked as active and it will not be used before the certificates are fetched. If Envoy fails to fetch the certificates due to connection failures, or bad response data, the cluster will be marked as active, it can be used to handle the requests, but the requests routed to that cluster will be rejected.
+
+If a static cluster is using SDS, and it needs to define a SDS cluster (unless google grpc is used which doens't need a cluster), the SDS cluster has to be defined before the static cluster using it.
```

Comment:
Hmm, I'm a bit confused here. Are you talking about literal bootstrap cluster order? This seems a bit surprising to me that we have a definition order dependency here. In ClusterManager, we already have an notion of primary/secondary initialization order to deal with EDS cluster that depend on the EDS delivery cluster, how come we didn't extend this notion to SDS? CC @mattklein123

Reply:
The current implementation of primary & secondary cluster initialization is simple;  only EDS is secondary, all other cluster types are primary.   In order to use this mechanism for SDS static clusters initialization, we have to mark any clusters using remote SDS as secondary.  

It is a TODO: created https://github.com/envoyproxy/envoy/issues/4522 to track this issue

Reply:
@qiwzhang thanks for the explanation.